### PR TITLE
CV Embeddable

### PR DESCRIPTION
**Pull request by acrap5**
Creates an embeddable CV carousel at /community-voices/public/embeddable.php. Filters can be applied to what slides appear by creating URLs in the same way one would search for slides on the dedicated page. Content category buttons can be toggled by optionally including `#buttons` in the URL.

https://www.notion.so/Create-CV-embeddable-db4dbb9dc44c414083d723e0610f01ec

### CODE REVIEW

File: src/CommunityVoices/Public/embeddable.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    function setCategory(category) {
+      window.location.search = `content_category[]=${category.dataset.cc}`
+    }
+    $(document).ready(function(){
+      if (window.location.hash !== "#buttons") {
+        document.getElementById("buttons").style.display = "none";
+      }
+    for(var i=1 ; i< paths.length ; i++) {
+      $('<div class="carousel-item"><div class="embed-responsive embed-responsive-16by9 mb-4"><iframe class="embed-responsive-item" id="slide2" style="pointer-events: none;" src="' + paths[i]+ '"></iframe></div></div>').appendTo('.carousel-inner');
+    }
```
[the remark this answers is not in the document]
Comment (Sammidysam) [reply src/CommunityVoices/Public/embeddable.php:160]: Though, leaving the HTML processing in JavaScript is good for dynamic updates of the page later on.

File: src/CommunityVoices/Public/embeddable.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      </div>
+    </div>
+
+    <script src="https://code.jquery.com/jquery-3.2.1.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>
+
+    <script>
+    var paths = <?php echo json_encode($files); ?>;
+    function setCategory(category) {
+      window.location.search = `content_category[]=${category.dataset.cc}`
```
Comment (Sammidysam) [inline src/CommunityVoices/Public/embeddable.php:152]: But does this reload the content on the page?

**Review by Sammidysam — COMMENTED** [review]
I worry that this setup does not make the category buttons truly filter, but I am not certain of that and need to give running this a go.

**Review by Sammidysam — CHANGES_REQUESTED** [review]
You may want to reconsider loading all slides into the carousel with no limits on how many slides you are selecting.  My browser was tending toward crashing quickly with loading what seemed like 2500 HTML nodes newly onto the page using JavaScript.

I'd consider using the API to load them one page at a time. https://environmentaldashboard.org/community-voices/api/slides
It is convenient that the API is paginated - can load page 0 then on slide 6 or so load page 1 into the HTML, and so on.

I figure you haven't loaded content through JavaScript like this as much (Tim never did much either), so let me know if you need to talk through that.  I just worry about performance here.

File: src/CommunityVoices/Public/embeddable.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
+<?php
+//TODO: Select approved only slides
```
[the remark this answers is not in the document]
Comment (acrap5) [reply src/CommunityVoices/Public/embeddable.php:2]: MVP? Anyways this is a comment I must've made at one point and just forgotten to get rid of.

[the remark this answers is not in the document]
Comment (Sammidysam) [reply src/CommunityVoices/Public/embeddable.php:2]: Minimum viable product, sorry.  The least you can do to get something done, even if you are going to improve it later.

[the remark this answers is not in the document]
Comment (acrap5) [reply src/CommunityVoices/Public/embeddable.php:2]: Right, regardless it was addressed, just forgot to take out my comment

File: src/CommunityVoices/Public/embeddable.php [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+$sql = 'SELECT probability, content_category_id, media_id, quote_id, image_id FROM `community-voices_slides` WHERE probability > 0 ' . $content_category_query . $search_query . $tag_query . $attribution_query . $photographer_query . $org_query . ' ORDER BY probability DESC';
+
+$weight_sum = array_sum($galleries);
+$sorted_rows = array_fill_keys($gallery_names, []); // list of urls, each duplicated to match its prob/weight
+$num_urls = 0;
+foreach ($dbHandler->query($sql) as $row) {
+    for ($i=0; $i < $row['probability']; $i++) {
+        $sorted_rows[$row['content_category_id']][] = "/community-voices/slides/{$row['media_id']}";
+    }-
```
[the remark this answers is not in the document]
Comment (acrap5) [reply src/CommunityVoices/Public/embeddable.php:70]: Actually didn't seem to be causing any issues, but shouldn't be there regardless, thanks for the catch.

File: src/CommunityVoices/Public/embeddable.php (hunk 2 above)
Comment (acrap5) [reply src/CommunityVoices/Public/embeddable.php:152]: Yes.

Comment (Sammidysam) [reply src/CommunityVoices/Public/embeddable.php:152]: Ahhh, it does a refresh.  I thought John / Rowan were envisioning a live update.

**Review by Sammidysam — APPROVED** [review]
Your solution is okay for a quick fix for making this not slow (John does want it imminently), but sometimes the event handler won't fire if someone is quickly pressing next over and over.  That could be kind of annoying.  But I would say let's get this live then we can talk through a design that will actually be nice?

**Comment by Sammidysam** [comment]
Feel free to merge and get live and then go ahead and email John and Rowan with a link.  Then maybe find time to do a brief call with me where we can plan out a way of doing this that might be better?

**Comment by Sammidysam** [comment]
One thing I realized after approving is that the buttons are not accessible without scrolling ... but I really think John gets into a bit too much of an atmosphere of it's not worth showing unless it's perfect, and that's really not the case here.  I think we should ship this version before fixing the height to incorporate buttons.
(For reference, it is very very hard to adjust the slide height with your current setup ... we can talk through how to get it to work better at some other time)

**Pull request merged**